src/verifast.ml

```diff
@@ -2897,6 +2910,7 @@ module VerifyProgram(VerifyProgramArgs: VERIFY_PROGRAM_ARGS) = struct
       execute_branch @@ fun () ->
       with_context (Executing ([], env, loc, sprintf "Verifying constructor '%s'" struct_name)) @@ fun () ->
       assume_neq this_term (null_pointer_term ()) @@ fun () ->
+      assume_eq (ctxt#mk_app cxx_typeid_symbol [this_term]) (List.assoc struct_name cxx_type_info_map) @@ fun () ->

```

Comment:
Could this not be a base ctor call? I.e. the most derived type of `this_term` might be a subtype of `struct_name`?

Reply:
§11.9.5 Construction and destruction [class.cdtor]/5: `(...) When typeid is used in a constructor (including the mem-initializer or default member initializer (11.4) for a non-static data member) or in a destructor, or used in a function called (directly or indirectly) from a constructor or destructor, if the operand of typeid refers to the object under construction or destruction, typeid yields the std::type_info object representing the constructor or destructor’s class. (...)
`

Reply:
Interesting! But this means that the type id of a given pointer changes during the lifetime of that pointer? So we cannot model the typeid as a pure function of the pointer? (Otherwise, I think you can prove false by stating typeid(this) == typeid(struct A) in the postcondition of the constructor for struct A; this will yield a contradiction when constructing a subclass object.

Reply:
But do we need `typeid(E)` yet in this PR? After all, there are no virtual methods yet, so there is no need yet for dynamic binding of instance predicates? Perhaps we should deal with expression typeids and with virtual methods in the same PR (because I suspect similar mechanisms will be necessary), and in this PR we do not yet support typeid(E).

Reply:
Makes sense. I'll remove support in this PR for `typeid(E)` where `E` is not a type expression.

Reply:
In fact, does it make sense to have virtual instance predicates (= predicate families indexed by a typeid) in a non-virtual class (= a class without a vtable)? Would it not be cleaner and more C++-like to have non-virtual instance predicates by default and have virtual instance predicates only if marked by keyword `virtual`?

Reply:
In Java, instance predicates are always virtual because methods are always virtual.

Reply:
At this point, I think it makes most sense to put this PR on hold and do the virtual member functions PR first.

Reply:
Do you also need the functionality to evaluate a `typeid(struct S)` expression to a `S_type_info` term?

Reply:
Yes, I do. Thanks!